storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3109,1760 @@ buf_block_t *fsp_get_latched_xdes_page(
+    if (err == DB_SUCCESS)
+      err= flst_validate(header, FSP_HEADER_OFFSET + FSP_SEG_INODES_FREE, &mtr);
+  }
+  mtr.commit();
+  return err;
+}
+
+dberr_t indexDefragmenter::store_child_pages(buf_block_t *block) noexcept
+{
+  const byte *page= block->page.frame;
+  const rec_t *rec= page_rec_get_next_low(page + PAGE_OLD_INFIMUM, false);
+  while (rec != page + PAGE_OLD_SUPREMUM)
+  {
+    ulint len;
+    ulint offset= rec_get_nth_field_offs_old(rec,
+                                             rec_get_n_fields_old(rec) - 1,
+                                             &len);
+    if (len != 4)
+      return DB_CORRUPTION;
+
+    if (offset >= srv_page_size)
+      return DB_CORRUPTION;
+
+    const byte *field= rec + offset;
+    /* m_parent_pages[child_page_no] =
+       1st 32 bit to indicate offset in parent page
+       2nd 32 bit to indicate parent page number */
+    m_parent_pages[mach_read_from_4(field)]=
+      uint64_t(page_offset(field)) << 32 | block->page.id().page_no();
+    rec= page_rec_get_next_low(rec, false);
+  }
+  return DB_SUCCESS;
+}
+
+uint32_t indexDefragmenter::get_level_block(uint16_t level, dberr_t *err) noexcept
+{
+  mtr_t mtr;
+  mtr.start();
+  mtr_x_lock_index(m_index, &mtr);
+  uint32_t child_page_no= m_index->page;
+  while (1)
+  {
+    dberr_t err= DB_SUCCESS;
+    buf_block_t *block= buf_page_get_gen(page_id_t{0, child_page_no},
+                                         0, RW_S_LATCH, nullptr,
+                                         BUF_GET_POSSIBLY_FREED, &mtr,
+                                         &err);
+    if (!block)
+    {
+      child_page_no= FIL_NULL;
+      break;
+    }
+
+    page_t *page= buf_block_get_frame(block);
+    uint16_t cur_level= btr_page_get_level(page);
+    if (cur_level == level)
+      break;
+    const rec_t *rec= page_rec_get_next_low(page + PAGE_OLD_INFIMUM, false);
+    if (rec && rec != page + PAGE_OLD_SUPREMUM)
+    {
+      ulint len;
+      ulint offset=
+        rec_get_nth_field_offs_old(rec, rec_get_n_fields_old(rec) - 1,
+                                   &len);
+      ut_ad(len == 4);
+
+      ut_ad(offset < srv_page_size);
+      child_page_no= mach_read_from_4(rec + offset);
+    }
+    else
+    {
+      child_page_no= FIL_NULL;
+      err= DB_CORRUPTION;
+      break;
+    }
+    if (cur_level == level + 1)
+      break;
+  }
+  mtr.commit();
+  return child_page_no;
+}
+
+dberr_t indexDefragmenter::defragment_level(uint16_t level,
+                                            spaceDefragmenter *space_defrag) noexcept
+{
+  dberr_t err= DB_SUCCESS;
+  uint32_t cur_page_no= get_level_block(level, &err);
+  if (cur_page_no == FIL_NULL)
+    return err;
+
+  fil_space_t *space= fil_system.sys_space;
+  uint32_t extent_size= space_defrag->m_extent_size;
+  mtr_t mtr;
+  mtr.start();
+  mtr.x_lock_space(space);
+  mtr_x_lock_index(m_index, &mtr);
+
+  buf_block_t *block= buf_page_get_gen(page_id_t{0, cur_page_no},
```

Comment:
This would seem to invert the lock order. A comment about that would be good to have if we can’t avoid this. Usually the order is that `dict_index_t::lock` is acquired first, then the page latches, then the tablespace latch, and finally latches on blocks that cover the allocation metadata.

I would suggest to acquire `m_index->lock` outside `mtr` and explicitly releasing it at the end. There should be no need to release and re-acquire the index latch inside the loop.